sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/conditionalExpressions.scala

```diff
@@ -137,45 +137,50 @@ case class CaseWhen(branches: Seq[(Expression, Expression)], elseValue: Option[E
   }
 
   override def genCode(ctx: CodeGenContext, ev: GeneratedExpressionCode): String = {
-    val got = ctx.freshName("got")
+    val len = branches.length
 
-    val cases = branches.map { case (condition, value) =>
-      val cond = condition.gen(ctx)
-      val res = value.gen(ctx)
-      s"""
-        if (!$got) {
+    def genBranch(branchIndex: Int): String =
+      if (branchIndex >= len) {
+        if (elseValue.isDefined) {
+          val res = elseValue.get.gen(ctx)
+          s"""
+            ${res.code}
+            ${ev.isNull} = ${res.isNull};
+            ${ev.value} = ${res.value};
+          """
+        } else {
+          ""
+        }
+      } else {
+        val cond = branches(branchIndex)._1.gen(ctx)
+        val res = branches(branchIndex)._2.gen(ctx)
+
+        val currentBranch = s"""
           ${cond.code}
           if (!${cond.isNull} && ${cond.value}) {
-            $got = true;
             ${res.code}
             ${ev.isNull} = ${res.isNull};
             ${ev.value} = ${res.value};
           }
-        }
-      """
-    }.mkString("\n")
-
-    val elseCase = {
-      if (elseValue.isDefined) {
-        val res = elseValue.get.gen(ctx)
-        s"""
-        if (!$got) {
-          ${res.code}
-          ${ev.isNull} = ${res.isNull};
-          ${ev.value} = ${res.value};
-        }
         """
-      } else {
-        ""
+
+        val moreBranch = genBranch(branchIndex + 1)
+        if (moreBranch != "") {
+          s"""
+            $currentBranch
+            else {
+              $moreBranch
```

Comment:
It will be slightly more efficient when we succeeds in front conditions and can skip checking of `got` variable for late branches. Before that we still need to check `got` content many times even first condition is already satisfied.